routes/xref/update.js

```diff
@@ -0,0 +1,81 @@
+const bodyParser = require("body-parser");
+const crypto = require("crypto");
+const { exec } = require("child_process");
+
+const bikeshedSecret = process.env.BIKESHED_SECRET;
+if (!bikeshedSecret) {
+  throw new ReferenceError("env variable `BIKESHED_SECRET` is not set.")
+}
+
+// minimum delay between subsequent updates
+const UPDATE_THRESHOLD = 60 * 60 * 1000; // 1 hour
+
+let lastUpdateTime = new Date(0);
+
+function validateGithubSignature(req, res, next) {
+  // see: https://developer.github.com/webhooks/securing/
+  const hash = crypto
+    .createHmac("sha1", bikeshedSecret)
+    .update(req.rawBody)
+    .digest("hex");
+
+  if (req.get("X-Hub-Signature") === `sha1=${hash}`) {
+    next();
+  } else {
+    next("Failed to authenticate GitHub hook Signature");
+  }
+}
+
+function ensureUpdateIsNeeded(req, res, next) {
+  if (req.body.refs !== "refs/heads/master") {
+    res.status(202); // Accepted
+    return res.send("Payload was not for master, aborted.");
+  }
+
+  if (!hasAnchorUpdate(req.body.commits)) {
+    res.status(202); // Accepted
+    return res.send("Anchors were not modified, aborted.");
+  }
+
+  if (new Date() - lastUpdateTime < UPDATE_THRESHOLD) {
+    res.status(429); // Too Many Requests
+    const waitUntil = new Date(lastUpdateTime.valueOf() + UPDATE_THRESHOLD);
+    const retryAfter = (waitUntil - new Date()).valueOf() / 1000; // in seconds
+    res.set("Retry-After", parseInt(retryAfter, 10));
+    return res.send("Tried to update too soon.");
+  }
+
+  next();
+}
+
+function handleBikeshedUpdate(req, res) {

```

Comment:
What about something like this:
```JS
const pendingExecs = [];

async function handleBikeshedUpdate(req, res, next) {
  lastUpdateTime = new Date();
  const promiseToExecute = new Promise(r => {
    exec("npm run get-xref-data", error => {
      if (error) {
        console.error("X-GitHub-Delivery", req.get("X-GitHub-Delivery"));
        res.status(500); // Internal Server Error
        res.send("Error while updating data. See server logs.");
        reject(error);
      }
      res.send("Updated.");
      resolve();
    });
  });
  pendingExecs.push(promiseToExecute);
  try {
    await Promise.all(pendingExecs);
  } catch(error) {
    console.error(error);
  } finally {
    next();
  }
}
```

Reply:
1. So, do we make use of `lastUpdateTime`? 
2. We are still executing it in same way, it's just wrapped in a Promise. I'm confused what we are trying to do.

Reply:
> lastUpdateTime

No, I don’t think we need it. 

> I'm confused what we are trying to do.

Heh, I shouldn’t make silly proposals without testing them. What we want is to await any current running process to finish... so, don’t want an array there, just a module variable that holds a promise representing the current thing executing.

Reply:
Ok, so was looking at this problem a bit more... we basically need a promise queue. I think we need it, but we should do it as a new PR, because it's none-trivial to implement. 

Also, checking locally, take note that PR has windows line endings again.